SDL-0234 Proxy Library RPC Generation (#1556)
smartdevicelink/sdl_ios: generator/README.md

# Proxy Library RPC Generator

## Overview

> This script provides a possibility to auto-generate Objective-C code (header \*.h and implementation \*.m classes) based on the SDL MOBILE_API XML specification provided as a source of true.

Reviewer: done

## Requirements

> The script requires **Python 3.5** pre-installed on the host system. This is the minimal **Python 3** version that has not yet reached the end-of-life (https://devguide.python.org/devcycle/#end-of-life-branches).

Reviewer: the code is compatible with 3.5 thus minimum required version is 3.5. This part just notifies that the generator will not work with previous versions. Even after 3.5 will reach EoL, the code will be compatible with that version until somebody puts the incompatible code. Do we still need this fix here?

> The script requires **Python 3.5** pre-installed on the host system. This is the minimal **Python 3** version that has not yet reached the end-of-life (https://devguide.python.org/devcycle/#end-of-life-branches).

Reviewer 2: I think the issue that I see here is that if the minimum required version is 3.5, we are _locked_ to that version until a major release of the library. Because 3.5 is close to EOL, bugfixes will not be created for it, and any language features newer than 3.5 cannot be used. In addition, I see no reason to choose 3.5 over a newer, but stable, release like 3.7.6. That gives us additional time to receive bug fixes without making a major version change to the library, and additional language features that can be used in the future if needed.

> The script requires **Python 3.5** pre-installed on the host system. This is the minimal **Python 3** version that has not yet reached the end-of-life (https://devguide.python.org/devcycle/#end-of-life-branches).

Author: @joeljfischer could the addition "3.5 or higher" fix that `lock` problem? To be honest, I don't think the readme really could lock maintainers to use features of the newest Python version if needed, however, this generator is 3.5 compatible by the fact at the moment.

> The script requires **Python 3.5** pre-installed on the host system. This is the minimal **Python 3** version that has not yet reached the end-of-life (https://devguide.python.org/devcycle/#end-of-life-branches).

Reviewer 2: I understand what you're saying. What I'm concerned about is if we say 3.5 or higher, then in a later release we use a version 3.7 part of the language, or rely on a bug fixed in a supported version. In that case, we would have to bump the minimum version, which would break users with Python 3.5 installed, which would be a major version change. We could say something like, "Minimum supported version is 3.7.6. It may work on versions back to 3.5, but this is not supported and may break in the future."

> The script requires **Python 3.5** pre-installed on the host system. This is the minimal **Python 3** version that has not yet reached the end-of-life (https://devguide.python.org/devcycle/#end-of-life-branches).

Author: @joeljfischer I'm ok with the last suggestion. I will put it in the document. The last stable release of Python is 3.8.2, are you ok with this version in the document?

> The script requires **Python 3.5** pre-installed on the host system. This is the minimal **Python 3** version that has not yet reached the end-of-life (https://devguide.python.org/devcycle/#end-of-life-branches).

Author: ```suggestion
The script requires **Python 3** pre-installed on the host system. The minimal supported Python 3 version is **3.8.2**.  It may work on versions back to 3.5 (the minimal version that has not yet reached [the end-of-life](https://devguide.python.org/devcycle/#end-of-life-branches)), but this is not supported and may break in the future.
```

> The script requires **Python 3.5** pre-installed on the host system. This is the minimal **Python 3** version that has not yet reached the end-of-life (https://devguide.python.org/devcycle/#end-of-life-branches).

Reviewer 2: Bilal and I were talking about that, I think we're looking for something a little more stable than 3.8.2, since there have been a few quick patches since the 3.8.0 release. Bilal and I agreed on 3.7.6.

> The script requires **Python 3.5** pre-installed on the host system. This is the minimal **Python 3** version that has not yet reached the end-of-life (https://devguide.python.org/devcycle/#end-of-life-branches).

Author: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/pull/1556/commits/50a2d3ea9e7bd7f95cc8f693856ef7451085f736

> Note: two pyton versions can be installed on the system and to use the required third version one must use the **pip3** command instead of the commonly used **pip**.

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

All required libraries are listed in `requirements.txt` and should be pre-installed on the system prior to using the sript. Please use the following command to install the libraries:

```shell script

> $ pip3 install -r requirements.txt

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

```

> Please also make sure all git submodules are installed and up to date since the script uses the XML parser provided there.

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

```shell script
$ git submodule update --init --recursive
```

or

```shell script

> $ git submodule update --recursive

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

> $ git submodule update --recursive

Reviewer 2: This was not done

> $ git submodule update --recursive

Reviewer: done in https://github.com/smartdevicelink/sdl_ios/pull/1556/commits/bbee519174c64e93af6a5e5f75cbe6d05f26d480

```

> 

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

## Usage

**Usage example**

```shell script
$ cd sdl_ios
$ python3 generator/generator.py -xml generator/rpc_spec/MOBILE_API.xml -xsd generator/rpc_spec/MOBILE_API.xsd -d output_dir
```

> *Note: one may skip the first item python3 if the host system configured to run \*.py files on the Python-3 engine.*

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

> As a result the output_dir will have all the new generated files (at the time of this article writing as many as 716 files were produced).

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

**Detailed usage description (keys, options)**

> 

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

```

> usage: generator.py [-h] [-v] [-xml SOURCE_XML] [-xsd SOURCE_XSD]

Reviewer: changed to:
```python3
' [Y/n]:\t'
```
@joeljfischer done in https://github.com/smartdevicelink/sdl_ios/pull/1556/commits/abb19f31a42d7d41122109e8a4993b7a4fb266cb

                    [-d OUTPUT_DIRECTORY] [-t [TEMPLATES_DIRECTORY]]
                    [-r REGEX_PATTERN] [--verbose] [-e] [-s] [-m] [-y] [-n]

Proxy Library RPC Generator

optional arguments:
  -h, --help            show this help message and exit
  -v, --version         print the version and exit
  -xml SOURCE_XML, --source-xml SOURCE_XML, --input-file SOURCE_XML
                        should point to MOBILE_API.xml
  -xsd SOURCE_XSD, --source-xsd SOURCE_XSD

>   -d OUTPUT_DIRECTORY, --output-directory OUTPUT_DIRECTORY

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/pull/1556/commits/abb19f31a42d7d41122109e8a4993b7a4fb266cb

>   -d OUTPUT_DIRECTORY, --output-directory OUTPUT_DIRECTORY

Reviewer 2: This was not fixed

                        define the place where the generated output should be
                        placed
  -t [TEMPLATES_DIRECTORY], --templates-directory [TEMPLATES_DIRECTORY]
                        path to directory with templates
  -r REGEX_PATTERN, --regex-pattern REGEX_PATTERN
                        only elements matched with defined regex pattern will
                        be parsed and generated
  --verbose             display additional details like logs etc
  -e, --enums           only specified elements will be generated, if present
  -s, --structs         only specified elements will be generated, if present

>   -m, -f, --functions   only specified elements will be generated, if present

Reviewer: this _Command-line switches_ was agreed in proposal:
https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0234-proxy-rpc-generation.md#command-line-switches
do you have other vision, what arguments should have generator?
Meanwhile to generate only particular element you ran use regex attribute as follow:
`-r AppHMIType`

>   -m, -f, --functions   only specified elements will be generated, if present

Reviewer 2: I understand the command line switches were agreed to in the proposal. I'm saying that I had a different interpretation of what the proposal is saying. In my mind, "only specified elements will be generated, if present" means the user can specify which elements they want to be generated.

I now understand that the interpretation of the proposal is "all enums / functions / structs will be generated." That's fine, but the README documentation can be clearer. I would recommend putting "all enums / functions / structs will be generated" as the switch documentation.

>   -m, -f, --functions   only specified elements will be generated, if present

Reviewer: done in https://github.com/smartdevicelink/sdl_ios/pull/1556/commits/4676ebdfec933116d3d47e8103da2677d1561675

  * Enums
  * Structs
  * Notification
  * Requests

>   * Responses

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

Due to the explosion of complexity and code that is involved with mapping every exception that has ever been made in the iOS project compared to the RPC spec, the decision was made that the generator should not pursue those mappings. This has several implications:

  * The generator will not override the existing RPCs. All newly created structs, enums, requests, responses, and notifications will be generated, but new parameters to existing RPCs will have to be manually handled. The script's `--skip` command line switch will be used.
  * Existing unit tests cannot be used to verify the output.

The idea is that a future major version could involve a switch-over to using all generated code instead of the current code. It is not perfect for the current version but it is helpful for adding new RPCs.

> Any major version change cannot be done that would be involved with overwriting existing RPC classes and to map out all the differences would be a huge time and complexity-sink that the PM and PR author agreed was feasible but unmaintainable and undesirable.

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

> #  Objetive-C transformation rules

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

> ## Overview

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

> These are the general transformation rules for SDL RPC classes Objective-C Library. The base classes description already included in the library is not provided here for more one may want to view the source code.

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

## Output Directory Structure and Package definitions

> The script creates corresponding RPC classes of `<enum>`, `<struct>` and `<function>` elements following the `MOBILE_API.xml` rules. According to existing structure of SmartDeviceLink library the output directory will contain the following files (plain structure, no subfolders). For enum, struct, function future naming, generally it should be named 

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

```SDL<Enum Name><Value>```

>  in camel case.

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

**Requests:**

  * SDLxxx.h
  * SDLxxx.m
  
**Responses:**

  * SDLxxxResponse.h
  * SDLxxxResponse.m

*Note: as an exception from rules this kind has the suffix 'Response'*
 
**Structures:**

  * SDLxxx.h
  * SDLxxx.m

**Enums:**

  * SDLxxx.h
  * SDLxxx.m

**Notifications:**

  * SDLxxx.h
  * SDLxxx.m

Where the **xxx** is the correspondent item name.

> 

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

## The License Header

All files should begin with the license information.

```
/*

>  * Copyright (c) 2017 - [year], SmartDeviceLink Consortium, Inc.

Reviewer: @joeljfischer, replaced with content of file `sdl_ios/generator/templates/copyright.txt`
done in https://github.com/smartdevicelink/sdl_ios/pull/1556/commits/abb19f31a42d7d41122109e8a4993b7a4fb266cb

 * this list of conditions and the following
 * disclaimer in the documentation and/or other materials provided with the
 * distribution.
 *
 * Neither the name of the SmartDeviceLink Consortium Inc. nor the names of
 * its contributors may be used to endorse or promote products derived
 * from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
```

> Where `[year]` in the copyright line is the current (?) year.

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

*Note: for simplisity sake the header removed from Objective-C code snippets in the following sections of this document*

> 

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

### General rules for Objective-C classes

> 1. Default initializer in every class 

Reviewer: @joeljfischer , changed to:
`1. Default initializer applies only to Functions(Request / Response / Notification) classes`
done in https://github.com/smartdevicelink/sdl_ios/pull/1556/commits/abb19f31a42d7d41122109e8a4993b7a4fb266cb

```objc
    - (instancetype)init {

>         if ((self = [super initWithName:SDLRPCFunctionNameRegisterAppInterface])) {

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/pull/1556/commits/abb19f31a42d7d41122109e8a4993b7a4fb266cb

        }
        return self;
    }
```

> *Pease note the required double brackets in the if statement*

Reviewer: @yLeonid please reply to this

> *Pease note the required double brackets in the if statement*

Reviewer 3: Example:
0. if (x=y) {} // is not a comparison but an assignment 
we can write it as:
1. if (nil != (x=y)) {}
or
2. if ((x=y)) {}
The issue with the #0 is that depending on compiler settings it will produce either a warning in better case or an error in worst.
The #2 is the most common way to avoid compiler warnings/errors.

Steps to reproduce the warning (error):
1. create a default command line objc project in Xcode.
2. copy-paste the following code:
```
#import <Foundation/Foundation.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        int x = 0;
        int y = 10;

        if (x = y) {
            NSLog(@"non zero");
        } else {
            NSLog(@"zero");
        }
    }
    return 0;
}
```
3. compile the project.
4. observe the following warning:
```
main.m:16:15: warning: using the result of an assignment as a condition without parentheses [-Wparentheses]
        if (x = y) {
            ~~^~~
/Users/leonid/ZProjects/tmp/XXX_test/XXX_test/main.m:16:15: note: place parentheses around the assignment to silence this warning
        if (x = y) {
              ^
            (    )
/Users/leonid/ZProjects/tmp/XXX_test/XXX_test/main.m:16:15: note: use '==' to turn this assignment into an equality comparison
        if (x = y) {
              ^
              ==
1 warning generated.
```
5. open the project settings.
6. go to the section: "Apple Clang - Warning Policies"
7. set "Treat Warnings as Errors" to 'yes'
8. compile the project.
9. observe the result:
```
main.m:16:15: error: using the result of an assignment as a condition without parentheses [-Werror,-Wparentheses]
        if (x = y) {
            ~~^~~
/Users/leonid/ZProjects/tmp/XXX_test/XXX_test/main.m:16:15: note: place parentheses around the assignment to silence this warning
        if (x = y) {
              ^
            (    )
/Users/leonid/ZProjects/tmp/XXX_test/XXX_test/main.m:16:15: note: use '==' to turn this assignment into an equality comparison
        if (x = y) {
              ^
              ==
1 error generated.

```

Conclusion:
Since the sdl_ios is a public available sdk (project) and as soon as it is released it is out of hands. We have no opportunity to control the user's compiler settings and in this case the best experience (IMHO) would be avoid as many warnings not saying of errors as possible especially if it takes next to none effort on our side.

On the other hand it is really not required and anyone can suppress any warnings in the project settings and so can the user.

Thank you for understanding.

2. Initializer for mandatory params if there is/are any in XML (skipped if no <mandatory = true> params)

> 

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

3. Initializer for all params if there is/are any which is not mandatory in XML (skipped if no <mandatory = false> params)

> ## Scalars.

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

There are 4 type of scalar values declared in the SDL lib.  These are:
1. **SDLInt** - A declaration that this NSNumber contains an NSInteger.
0. **SDLUInt** - A declaration that this NSNumber contains an NSUInteger.
0. **SDLBool** - A declaration that this NSNumber contains a BOOL.
0. **SDLFloat** - A declaration that this NSNumber contains a float.

> *Note: the ones declared as empty protocols and NSNumber conforms to them.*

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

```objc
@interface NSNumber (NumberType) <SDLInt, SDLUInt, SDLBool, SDLFloat>
```
Usage example:
```objc
@property (strong, nonatomic) NSNumber<SDLInt> *touchEventId;
```
or an array:
```objc
@property (strong, nonatomic) NSArray<NSNumber<SDLInt> *> *timeStamp;
```

> *Note: the 4 scalar values implemented as empty protocols and 'extend' the* **NSNumber** *class.*

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

> ## Enums

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

> So called enums in **sdl\_ios** implemented as strings (NSString) typedefed with a proper enum type. One may notice it is not real enums but NSString objects though in SWIFT they must become real enums.

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

*Example:*

```objc
typedef NSString* SDLEnum SDL_SWIFT_ENUM;

> ```

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

> *Note: This new defined type has already had an asterisk at the end so anything that inherits from SDLEnum needs no asterisk.*

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

```objc

> typedef SDLEnum SDLTouchType SDL_SWIFT_ENUM;

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

```

> *Note: The compiler considers SDLTouchType as NSString**

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

> And here is a concrete 'enum' item

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

```objc
extern SDLTouchType const SDLTouchTypeBegin;
```

If an item is deprecated then it will be declared as such:

```objc
extern SDLTouchType const SDLTouchTypeBegin __deprecated;
```

or even:

```objc

> extern SDLTouchType const SDLTouchTypeBegin __deprecated_msg(("this item is deprecated once and for all, please use SDLTouchTypeNewType instead"));

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/pull/1556/commits/abb19f31a42d7d41122109e8a4993b7a4fb266cb

```

> 

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

Each Enum class is stored in two files (header \*.h and implementation \*.m)  and the filename and the class name should be composed based on the value from the `"name"` attribute of `<enum>`.

Example:

```
<enum name="ImageType" .../>
./SDLImageType.h
./SDLImageType.m
```

> 

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

Each Enum class should include the enum definition, see for example the AmbientLightStatus enum:

```objc
#import "SDLEnum.h"
typedef SDLEnum SDLAmbientLightStatus SDL_SWIFT_ENUM;
```

> 

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

Take for an instance the enum class KeypressMode
 
```xml
<enum name="KeypressMode" since="3.0">
    <description>Enumeration listing possible keyboard events.</description>
    <element name="SINGLE_KEYPRESS">
        <description>Each keypress is individually sent as the user presses the keyboard keys.</description>
    </element>
    <element name="QUEUE_KEYPRESSES">
        <description>The keypresses are queued and a string is eventually sent once the user chooses to submit their entry.</description>
    </element>
    <element name="RESEND_CURRENT_ENTRY">
        <description>The keypresses are queue and a string is sent each time the user presses a keyboard key; the string contains the entire current entry.</description>

>     </element>

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

</enum>
```

> In the following example 

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

```objc
extern SDLKeypressMode const SDLKeypressModeSingleKeypress;
```

The **SDLKeypressModeSingleKeypress** enum is seen by the compiler as:

```objc
extern NSString* const SDLKeypressModeSingleKeypress;
```

> 

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

> and **SDLKeypressModeSingleKeypress** itself must be implemented in the correspondent SDLKeypressMode.m file as a string:

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

```objc
SDLKeypressMode const SDLKeypressModeSingleKeypress = @"SINGLE_KEYPRESS";
```

> ## Structures

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

> Structures in **sdl_ios**  implemented as classes derived from the parent class SDLRPCStruct with all parameters implemented as "@property". Let us take for an instance the **DeviceInfo** structure. In the XML it is declared as following:

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

<struct name="DeviceInfo" since="3.0">
        <description>Various information about connecting device.</description>        
        <param name="hardware" type="String"  minlength="0" maxlength="500" mandatory="false">
            <description>Device model</description>
        </param>
        <param name="firmwareRev" type="String" minlength="0" maxlength="500" mandatory="false">
            <description>Device firmware revision</description>
        </param>
        <param name="os" type="String" minlength="0" maxlength="500" mandatory="false">
            <description>Device OS</description>
        </param>
        <param name="osVersion" type="String" minlength="0" maxlength="500" mandatory="false">
            <description>Device OS version</description>
        </param>
        <param name="carrier" type="String" minlength="0" maxlength="500" mandatory="false">
            <description>Device mobile carrier (if applicable)</description>
        </param>
        <param name="maxNumberRFCOMMPorts" type="Integer" minvalue="0" maxvalue="100" mandatory="false">
            <description>Omitted if connected not via BT.</description>
        </param>        
 </struct>
 ```
 

>  *Please note that all params declared as mandatory="false" and there is one init method with all the params in the generated file. The method* ```+ (instancetype)currentDevice;``` *comes from the old manually made implementation*

Reviewer 2: This is unnecessary since the generator won't generate it.

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/pull/1556/commits/abb19f31a42d7d41122109e8a4993b7a4fb266cb

>  *All method and property descriptions generated from the xml descriptions*

Reviewer 2: ```suggestion
```

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

>  *Note: the file begins with the* **NS_ASSUME_NONNULL_BEGIN** *macro which makes all properties / parameters mandatory. If a parameter is not mandatory then the modifier* **nullable** *will be used*

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

```objc

> //  SDLDeviceInfo.h

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/pull/1556/commits/abb19f31a42d7d41122109e8a4993b7a4fb266cb

/**
 * Device model
 * {"default_value": null, "max_length": 500, "min_length": 0}
 *
 * Optional, NSString *
 */
@property (nullable, strong, nonatomic) NSString *hardware;

/**
 * Device firmware revision
 * {"default_value": null, "max_length": 500, "min_length": 0}
 *
 * Optional, NSString *
 */
@property (nullable, strong, nonatomic) NSString *firmwareRev;

/**
 * Device OS
 * {"default_value": null, "max_length": 500, "min_length": 0}
 *
 * Optional, NSString *
 */
@property (nullable, strong, nonatomic) NSString *os;

/**
 * Device OS version
 * {"default_value": null, "max_length": 500, "min_length": 0}
 *
 * Optional, NSString *
 */
@property (nullable, strong, nonatomic) NSString *osVersion;

/**
 * Device mobile carrier (if applicable)
 * {"default_value": null, "max_length": 500, "min_length": 0}
 *
 * Optional, NSString *
 */
@property (nullable, strong, nonatomic) NSString *carrier;

/**
 * Omitted if connected not via BT.
 * {"default_value": null, "max_value": 100, "min_value": 0}
 *
 * Optional, UInt8
 */
@property (nullable, strong, nonatomic) NSNumber<SDLUInt> *maxNumberRFCOMMPorts;

@end

NS_ASSUME_NONNULL_END
```

The implementation **SDLDeviceInfo.m** file:

> ```objc

Reviewer: @joeljfischer done inhttps://github.com/smartdevicelink/sdl_ios/pull/1556/commits/abb19f31a42d7d41122109e8a4993b7a4fb266cb

}

- (void)setOs:(nullable NSString *)os {
    [self.store sdl_setObject:os forName:SDLRPCParameterNameOS];
}

- (nullable NSString *)os {
    return [self.store sdl_objectForName:SDLRPCParameterNameOS ofClass:NSString.class error:nil];
}

- (void)setOsVersion:(nullable NSString *)osVersion {
    [self.store sdl_setObject:osVersion forName:SDLRPCParameterNameOSVersion];
}

- (nullable NSString *)osVersion {
    return [self.store sdl_objectForName:SDLRPCParameterNameOSVersion ofClass:NSString.class error:nil];
}

- (void)setCarrier:(nullable NSString *)carrier {
    [self.store sdl_setObject:carrier forName:SDLRPCParameterNameCarrier];
}

- (nullable NSString *)carrier {
    return [self.store sdl_objectForName:SDLRPCParameterNameCarrier ofClass:NSString.class error:nil];
}

- (void)setMaxNumberRFCOMMPorts:(nullable NSNumber<SDLUInt> *)maxNumberRFCOMMPorts {
    [self.store sdl_setObject:maxNumberRFCOMMPorts forName:SDLRPCParameterNameMaxNumberRFCOMMPorts];
}

- (nullable NSNumber<SDLUInt> *)maxNumberRFCOMMPorts {
    return [self.store sdl_objectForName:SDLRPCParameterNameMaxNumberRFCOMMPorts ofClass:NSNumber.class error:nil];
}

@end

NS_ASSUME_NONNULL_END
```

> ## Functions

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

> Functions in iOS implemented as 3 different classes grouped by its kind, all the 3 extend the common parent class **SDLRPCMessage**

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

> ```objc

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

// This is the parent class for all the 3 function kinds

> SDLRPCMessage : SDLRPCStruct <NSCopying> : NSObject <NSCopying>

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

SDLRPCNotification : SDLRPCMessage  /// An RPC sent from the head unit to the app about some data change, such as a button was pressed
SDLRPCRequest : SDLRPCMessage   /// Superclass of RPC requests
SDLRPCResponse : SDLRPCMessage  /// Superclass of RPC responses
```

> *Note: for some reason SDLRPCMessage conforms to NSCopying protocol twice.*

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

> First of all there is the **FunctionID** class generated though it is not declared in the XML. This class maps all function IDs that are integers to function names as strings.

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

1. Uses of the `"name"` attribute should be normalized by the removal of the ID suffix, e.g. `RegisterAppInterfaceID -> RegisterAppInterface`. 

>   

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

> 0. The constant name should be camel case formatted.

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

>   

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

0. The constant has 2 fields the first is the `int` value of the `"value"` attribute and the second is the `String` value of normalized `"name"` attribute.

> Internally it uses another file that lists all the function names **SDLRPCFunctionNames**

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

```objc

> //  SDLFunctionID.h

Reviewer: The exact FunctionID has completely another structure that other enums, are you sure that it is good idea to put into documentation generated SDLFunctionID.h instead of existing one?

> //  SDLFunctionID.h

Reviewer 2: I agree.

> //  SDLFunctionID.h

Reviewer: done in https://github.com/smartdevicelink/sdl_ios/pull/1556/commits/45f8005b9c7bd32e72f8c0891e85feb55d688d93

#import <Foundation/Foundation.h>
#import "NSNumber+NumberType.h"
#import "SDLRPCFunctionNames.h"

NS_ASSUME_NONNULL_BEGIN

/// A function ID for an SDL RPC
@interface SDLFunctionID : NSObject

/// The shared object for pulling function id information
+ (instancetype)sharedInstance;

/// Gets the function name for a given SDL RPC function ID
///
/// @param functionID A function ID
/// @returns An SDLRPCFunctionName
- (nullable SDLRPCFunctionName)functionNameForId:(UInt32)functionID;

/// Gets the function ID for a given SDL RPC function name
///
/// @param functionName The RPC function name
- (nullable NSNumber<SDLInt> *)functionIdForName:(SDLRPCFunctionName)functionName;

@end

NS_ASSUME_NONNULL_END
```

> Each <function> from MOBILE_API.XML is declared in SDLRPCFunctionNames.h and SDLRPCFunctionNames.m files. It exports function names as strings.

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

```objc
SDLRPCFunctionNames.h
#import "SDLEnum.h"
/**
* All RPC request / response / notification names
*/
typedef SDLEnum SDLRPCFunctionName SDL_SWIFT_ENUM;

/// Function name for an AddCommand RPC
extern SDLRPCFunctionName const SDLRPCFunctionNameAddCommand;

/// Function name for an AddSubMenu RPC
extern SDLRPCFunctionName const SDLRPCFunctionNameAddSubMenu;

. . . and so on
```

And the implementation file SDLRPCFunctionNames.m :

```objc
//
//  SDLRPCFunctionNames.m
//  SmartDeviceLink
//

#import "SDLRPCFunctionNames.h"

SDLRPCFunctionName const SDLRPCFunctionNameAddCommand = @"AddCommand";
SDLRPCFunctionName const SDLRPCFunctionNameAddSubMenu = @"AddSubMenu";

. . . and so on
```

> 

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

> ### Request Functions (SDLRPCRequest)

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

```xml
<function name="ListFiles" functionID="ListFilesID" messagetype="request" since="3.0">
    <description>
         Requests the current list of resident filenames for the registered app.
         Not supported on first generation SDL enabled vehicles.
    </description>
</function>
```

Declaration file (.h)

```objc
//  SDLListFiles.h
//

#import "SDLRPCRequest.h"

NS_ASSUME_NONNULL_BEGIN

/**
 * Requests the current list of resident filenames for the registered app. Not supported on first generation SDL
 * enabled vehicles.
 *
 * @since SDL 3.0.0
 */
@interface SDLListFiles : SDLRPCRequest

@end

NS_ASSUME_NONNULL_END
```

Implementation file (.m)

```objc
//  SDLListFiles.m

#import "SDLListFiles.h"
#import "NSMutableDictionary+Store.h"
#import "SDLRPCFunctionNames.h"

NS_ASSUME_NONNULL_BEGIN

@implementation SDLListFiles

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
- (instancetype)init {
    if ((self = [super initWithName:SDLRPCFunctionNameListFiles])) {
    }
    return self;
}
#pragma clang diagnostic pop

@end

NS_ASSUME_NONNULL_END
```

> 

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

##### Function with simple params

This section depicts all functions which include one or a few parameters of the following types (integer|decimal|boolean|string). Such parameters are considered as simple.

```xml
    <function name="GetCloudAppProperties" functionID="GetCloudAppPropertiesID" messagetype="request" since="5.1">
        <description>
            RPC used to get the current properties of a cloud application
        </description> 
        <param name="appID" type="String" maxlength="100" mandatory="true"></param>
    </function>
```

### Request Functions (SDLRPCRequest)

> The parent class is **SDLRPCRequest**

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

 *
 * @since SDL 5.1.0
 */
@interface SDLGetCloudAppProperties : SDLRPCRequest

/**
 * @param appID
 * @return A SDLGetCloudAppProperties object
 */
- (instancetype)initWithAppID:(NSString *)appID;

/**
 * {"default_value": null, "max_length": 100, "min_length": 1}
 *
 * Required, NSString *
 */
@property (strong, nonatomic) NSString *appID;

@end

NS_ASSUME_NONNULL_END
```

```objc
//  SDLGetCloudAppProperties.m
//

#import "SDLGetCloudAppProperties.h"
#import "NSMutableDictionary+Store.h"
#import "SDLRPCFunctionNames.h"

NS_ASSUME_NONNULL_BEGIN

@implementation SDLGetCloudAppProperties

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
- (instancetype)init {
    if ((self = [super initWithName:SDLRPCFunctionNameGetCloudAppProperties])) {
    }
    return self;
}
#pragma clang diagnostic pop

- (void)setAppID:(NSString *)appID {
    [self.parameters sdl_setObject:appID forName:SDLRPCParameterNameAppId];
}

- (NSString *)appID {
    NSError *error = nil;
    return [self.parameters sdl_objectForName:SDLRPCParameterNameAppId ofClass:NSString.class error:&error];
}

@end

NS_ASSUME_NONNULL_END
```

> ### Response Functions (SDLRPCResponse)

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

NS_ASSUME_NONNULL_BEGIN

@implementation SDLAddCommandResponse

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
- (instancetype)init {
    if ((self = [super initWithName:SDLRPCFunctionNameAddCommand])) {
    }
    return self;
}
#pragma clang diagnostic pop

@end

NS_ASSUME_NONNULL_END
```

As one may notice this class does not implement too much since the main logic is already implemented in its parent class:

```objc
//  SDLRPCResponse.h
//

#import "SDLRPCMessage.h"
#import "SDLResult.h"

NS_ASSUME_NONNULL_BEGIN

/// Superclass of RPC responses
@interface SDLRPCResponse : SDLRPCMessage

/**
 *  The correlation id of the corresponding SDLRPCRequest.
 */
@property (strong, nonatomic) NSNumber<SDLInt> *correlationID;

/**
 *  Whether or not the SDLRPCRequest was successful.
 */
@property (strong, nonatomic) NSNumber<SDLBool> *success;

/**
 *  The result of the SDLRPCRequest. If the request failed, the result code contains the failure reason.
 */
@property (strong, nonatomic) SDLResult resultCode;

/**
 *  More detailed success or error message.
 */
@property (nullable, strong, nonatomic) NSString *info;

@end

NS_ASSUME_NONNULL_END
```

> 

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

> #### Response class with extra argument example (Alert Response):

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

            <element name="TOO_MANY_PENDING_REQUESTS"/>
            <element name="APPLICATION_NOT_REGISTERED"/>
            <element name="GENERIC_ERROR"/>
            <element name="REJECTED"/>
            <element name="ABORTED"/>
            <element name="DISALLOWED"/>
            <element name="USER_DISALLOWED"/>
            <element name="UNSUPPORTED_RESOURCE"/>
            <element name="WARNINGS"/>
        </param>
        
        <param name="info" type="String" maxlength="1000" mandatory="false" platform="documentation">
            <description>Provides additional human readable info regarding the result.</description>
        </param>
        
        <param name="tryAgainTime" type="Integer" minvalue="0" maxvalue="2000000000" mandatory="false" since="2.0">
            <description>
                Amount of time (in seconds) that an app must wait before resending an alert.
                If provided, another system event or overlay currently has a higher priority than this alert.
                An app must not send an alert without waiting at least the amount of time dictated.
            </description>
        </param>
        

>     </function>

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

```

> As we can see it in the XML it requites an additional param **tryAgainTime** which is declared and implemented in the **SDLAlertResponse** class as follows:

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

 * overlay currently has a higher priority than this alert. An app must not send an alert without waiting at least
 * the amount of time dictated.
 * {"default_value": null, "max_value": 2000000000, "min_value": 0}
 *
 * @since SDL 2.0.0
 *
 * Optional, UInt32
 */
@property (nullable, strong, nonatomic) NSNumber<SDLUInt> *tryAgainTime;

@end

NS_ASSUME_NONNULL_END
```

```objc
//  SDLAlertResponse.m
//

#import "SDLAlertResponse.h"
#import "NSMutableDictionary+Store.h"
#import "SDLRPCFunctionNames.h"

NS_ASSUME_NONNULL_BEGIN

@implementation SDLAlertResponse

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
- (instancetype)init {
    if ((self = [super initWithName:SDLRPCFunctionNameAlert])) {
    }
    return self;
}
#pragma clang diagnostic pop

- (void)setTryAgainTime:(nullable NSNumber<SDLUInt> *)tryAgainTime {
    [self.parameters sdl_setObject:tryAgainTime forName:SDLRPCParameterNameTryAgainTime];
}

- (nullable NSNumber<SDLUInt> *)tryAgainTime {
    return [self.parameters sdl_objectForName:SDLRPCParameterNameTryAgainTime ofClass:NSNumber.class error:nil];
}

@end

NS_ASSUME_NONNULL_END
```

> ### Notification Functions (SDLRPCNotification)

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

The parent class is **SDLRPCNotification**

> 

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

```xml
<function name="OnAudioPassThru" functionID="OnAudioPassThruID" messagetype="notification" since="2.0">
        <description>Binary data is in binary part of hybrid msg</description>
    </function>
```

The corresponding Objective-C class **SDLOnAudioPassThru** looks as following.

```objc
//  SDLOnAudioPassThru.h

#import "SDLRPCNotification.h"

NS_ASSUME_NONNULL_BEGIN

/**
 * Binary data is in binary part of hybrid msg
 *
 * @since SDL 2.0.0
 */
@interface SDLOnAudioPassThru : SDLRPCNotification

@end

NS_ASSUME_NONNULL_END
```

```objc
//  SDLOnAudioPassThru.m

#import "SDLOnAudioPassThru.h"
#import "NSMutableDictionary+Store.h"
#import "SDLRPCParameterNames.h"
#import "SDLRPCFunctionNames.h"

NS_ASSUME_NONNULL_BEGIN

@implementation SDLOnAudioPassThru

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
- (instancetype)init {
    if ((self = [super initWithName:SDLRPCFunctionNameOnAudioPassThru])) {
    }
    return self;
}
#pragma clang diagnostic pop

@end

NS_ASSUME_NONNULL_END
```

> Another example is the class **SDLOnAppInterfaceUnregistered**. It is declared in the XML as follows.

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

```xml
<function name="OnAppInterfaceUnregistered" functionID="OnAppInterfaceUnregisteredID" messagetype="notification" since="1.0">
        <param name="reason" type="AppInterfaceUnregisteredReason" mandatory="true">
            <description>See AppInterfaceUnregisteredReason</description>
        </param>

>     </function>

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

```

> The corresponding Objective-C class **SDLOnAppInterfaceUnregistered** looks as following.

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

```objc
//  SDLOnAppInterfaceUnregistered.h

#import "SDLRPCNotification.h"

@class SDLAppInterfaceUnregisteredReason;

NS_ASSUME_NONNULL_BEGIN

/**
 * @since SDL 1.0.0
 */
@interface SDLOnAppInterfaceUnregistered : SDLRPCNotification

/**
 * @param reason - @(reason)
 * @return A SDLOnAppInterfaceUnregistered object
 */
- (instancetype)initWithReason:(SDLAppInterfaceUnregisteredReason)reason;

/**
 * See AppInterfaceUnregisteredReason
 *
 * Required, SDLAppInterfaceUnregisteredReason 
 */
@property (strong, nonatomic) SDLAppInterfaceUnregisteredReason reason;

@end

NS_ASSUME_NONNULL_END
```

> *Note: though the `SDLAppInterfaceUnregisteredReason reason` declared as a property but implemented as custom getter and setter*

Reviewer: @yLeonid could you clarify

> *Note: though the `SDLAppInterfaceUnregisteredReason reason` declared as a property but implemented as custom getter and setter*

Reviewer 3: it is @dynamic, not @synthesize and as such the property requires a getter and a setter implementation

> *Note: though the `SDLAppInterfaceUnregisteredReason reason` declared as a property but implemented as custom getter and setter*

Reviewer 2: That's true of all RPC properties. I'm going to recommend removal.

> *Note: though the `SDLAppInterfaceUnregisteredReason reason` declared as a property but implemented as custom getter and setter*

Reviewer: @joeljfischer any activities required from our side here?

> *Note: though the `SDLAppInterfaceUnregisteredReason reason` declared as a property but implemented as custom getter and setter*

Reviewer 2: I added a comment on the newest review, if that is done, then no other action is necessary.

```objc
//  SDLOnAppInterfaceUnregistered.m

#import "SDLOnAppInterfaceUnregistered.h"
#import "NSMutableDictionary+Store.h"
#import "SDLAppInterfaceUnregisteredReason.h"
#import "SDLRPCParameterNames.h"
#import "SDLRPCFunctionNames.h"

NS_ASSUME_NONNULL_BEGIN

@implementation SDLOnAppInterfaceUnregistered

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
- (instancetype)init {
    if ((self = [super initWithName:SDLRPCFunctionNameOnAppInterfaceUnregistered])) {
    }
    return self;
}
#pragma clang diagnostic pop

- (instancetype)initWithReason:(SDLAppInterfaceUnregisteredReason)reason {
    self = [super init];
    if (!self) {
        return nil;
    }
    self.reason = @(reason);
    return self;
}

- (void)setReason:(SDLAppInterfaceUnregisteredReason)reason {
    [self.parameters sdl_setObject:reason forName:SDLRPCParameterNameReason];
}

- (SDLAppInterfaceUnregisteredReason)reason {
    NSError *error = nil;
    return [self.parameters sdl_enumForName:SDLRPCParameterNameReason error:&error];
}

@end

NS_ASSUME_NONNULL_END
```

> ## Utils

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

### Generator

> Proxy Library RPC Generator inherits any license defined in SDL BASE.

Reviewer: @joeljfischer done in https://github.com/smartdevicelink/sdl_ios/commit/3c6cd6d16aef99044381a6b869abe6f077c4fc04

 * this list of conditions and the following
 * disclaimer in the documentation and/or other materials provided with the
 * distribution.
 *
 * Neither the name of the SmartDeviceLink Consortium Inc. nor the names of
 * its contributors may be used to endorse or promote products derived
 * from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
```

> Where `[year]` in the copyright line is the current (?) year.

Reviewer 2: ```suggestion
Where `{{year}}` in the copyright line is the current year.
```

### General rules for Objective-C classes
1. Default initializer applies only to Functions(Request / Response / Notification) classes

```objc
    - (instancetype)init {
        self = [super initWithName:SDLRPCFunctionNameRegisterAppInterface];
        if (!self) { return nil; }

        return self;
    }
```
*Pease note the required double brackets in the if statement*

2. Initializer for mandatory params if there is/are any in XML (skipped if no <mandatory = true> params)
3. Initializer for all params if there is/are any which is not mandatory in XML (skipped if no <mandatory = false> params)

#### Scalars
There are 4 type of scalar values declared in the SDL lib.  These are:
1. **SDLInt** - A declaration that this NSNumber contains an NSInteger.
0. **SDLUInt** - A declaration that this NSNumber contains an NSUInteger.
0. **SDLBool** - A declaration that this NSNumber contains a BOOL.
0. **SDLFloat** - A declaration that this NSNumber contains a float.

> *Note: These are syntactic sugar to help the developer know what type of value is held in the `NSNumber`.*

Reviewer 2: ```suggestion

*Note: These are syntactic sugar to help the developer know what type of value is held in the `NSNumber`.*
```

```objc
@interface NSNumber (NumberType) <SDLInt, SDLUInt, SDLBool, SDLFloat>

> ```

Reviewer 2: ```suggestion
```

Usage example:
```objc
@property (strong, nonatomic) NSNumber<SDLInt> *touchEventId;
```

> or an array:

Reviewer 2: ```suggestion

or in an array:
```

```objc
@property (strong, nonatomic) NSArray<NSNumber<SDLInt> *> *timeStamp;
```

#### Enums
RPC Enums in SDL are strings. sdl_ios uses `NSString` `typedef`ed with a proper enum type. In Swift projects, however, they become real enums by using the `NS_SWIFT_ENUM` compiler tag.

> *Base definition of `SDLEnum`:*

Reviewer 2: ```suggestion
Base definition of `SDLEnum`:
```

```objc
typedef NSString* SDLEnum SDL_SWIFT_ENUM;

*Note: This new defined type has already adds a pointer, so anything that inherits from `SDLEnum` needs no asterisk.*

```objc
typedef SDLEnum SDLTouchType SDL_SWIFT_ENUM; // SDLTouchType will be considered an NSString by the compiler in Obj-C, but will be an enum object of type SDLTouchType in Swift.
```

And here is a concrete 'enum' item:

```objc
extern SDLTouchType const SDLTouchTypeBegin;
```

If an item is deprecated then it will be declared as such:

```objc
__deprecated

> extern SDLTouchType const SDLTouchTypeBegin;

Reviewer 2: This is not show it should be generated. It should be generated as:

```objc
extern SDLTouchType const SDLTouchTypeBegin __deprecated;
```

Reviewer: @joeljfischer please have look into commit below, if understand you right  https://github.com/smartdevicelink/sdl_ios/pull/1556/commits/0f53ab92a5589ee857aa96104f6f7c34e319e4fc

Reviewer 2: I don't think that that's correct. If I'm understanding this code change correctly, you're changing the `@interface` definition to have `__deprecated` next to it instead of above it. That's not correct and will cause a build error. For `@interface` definitions, it should be above it.

Not:

```objc
@interface SDLRPCMessage : SDLRPCStruct <NSCopying> __deprecated
```

Instead:

```objc
__deprecated
@interface SDLRPCMessage : SDLRPCStruct <NSCopying>
```

However, for parameters or for constants (like the one referenced in the initial comment) it should be next to it like I described in my first comment. Does that make sense?

Reviewer 2: Basically it should be above a type definition and next to a parameter or constant. I can't currently think of any other cases.

Reviewer: done in https://github.com/smartdevicelink/sdl_ios/pull/1556/commits/45f8005b9c7bd32e72f8c0891e85feb55d688d93

}

- (void)setOs:(nullable NSString *)os {
    [self.store sdl_setObject:os forName:SDLRPCParameterNameOs];
}

- (nullable NSString *)os {
    return [self.store sdl_objectForName:SDLRPCParameterNameOs ofClass:NSString.class error:nil];
}

- (void)setOsVersion:(nullable NSString *)osVersion {
    [self.store sdl_setObject:osVersion forName:SDLRPCParameterNameOsVersion];
}

- (nullable NSString *)osVersion {
    return [self.store sdl_objectForName:SDLRPCParameterNameOsVersion ofClass:NSString.class error:nil];
}

- (void)setCarrier:(nullable NSString *)carrier {
    [self.store sdl_setObject:carrier forName:SDLRPCParameterNameCarrier];
}

- (nullable NSString *)carrier {
    return [self.store sdl_objectForName:SDLRPCParameterNameCarrier ofClass:NSString.class error:nil];
}

- (void)setMaxNumberRFCOMMPorts:(nullable NSNumber<SDLUInt> *)maxNumberRFCOMMPorts {
    [self.store sdl_setObject:maxNumberRFCOMMPorts forName:SDLRPCParameterNameMaxNumberRFCOMMPorts];
}

- (nullable NSNumber<SDLUInt> *)maxNumberRFCOMMPorts {
    return [self.store sdl_objectForName:SDLRPCParameterNameMaxNumberRFCOMMPorts ofClass:NSNumber.class error:nil];
}

@end

NS_ASSUME_NONNULL_END
```

#### Functions

> Functions in iOS are implemented as 3 different classes grouped by their respective type. All the 3 extend the common parent class `SDLRPCMessage`.

Reviewer 2: ```suggestion
Functions in iOS are implemented as 3 different classes (`SDLRPCRequest`, `SDLRPCResponse`, and `SDLRPCNotification`) grouped by their respective type. All the 3 extend the common parent class `SDLRPCMessage`.
```

// This is the parent class for all the 3 function kinds
SDLRPCMessage <NSCopying> : SDLRPCStruct <NSCopying> : NSObject

SDLRPCNotification : SDLRPCMessage  /// An RPC sent from the head unit to the app about some data change, such as a button was pressed
SDLRPCRequest : SDLRPCMessage   /// Superclass of RPC requests
SDLRPCResponse : SDLRPCMessage  /// Superclass of RPC responses

> ```

Reviewer 2: ```suggestion
```

> First of all there is the `SDLFunctionID` class generated though it is not declared in the XML. This class maps all function IDs that are integers to function names as strings.

Reviewer 2: ```suggestion
##### Function ID, Function Name, and Parameter Name Special Case Class
There is also the `SDLFunctionID` class generated though it is not declared in the XML. This class maps all function IDs that are integers to function names as strings.
```

1. Uses of the `"name"` attribute should be normalized by the removal of the ID suffix, e.g. `RegisterAppInterfaceID -> RegisterAppInterface`. 
2. The constant name should be camel case formatted.

> 0. The constant has 2 fields the first is the `int` value of the `"value"` attribute and the second is the `String` value of normalized `"name"` attribute.

Reviewer 2: ```suggestion
3. The constant has 2 fields the first is the `int` value of the `"value"` attribute and the second is the `String` value of normalized `"name"` attribute.
```

Internally it uses another file that lists all the function names `SDLRPCFunctionNames`.

```objc
//  SDLFunctionID.h

#import <Foundation/Foundation.h>
#import "NSNumber+NumberType.h"
#import "SDLRPCFunctionNames.h"

NS_ASSUME_NONNULL_BEGIN

/// A function ID for an SDL RPC
@interface SDLFunctionID : NSObject

/// The shared object for pulling function id information
+ (instancetype)sharedInstance;

/// Gets the function name for a given SDL RPC function ID
///
/// @param functionID A function ID
/// @returns An SDLRPCFunctionName
- (nullable SDLRPCFunctionName)functionNameForId:(UInt32)functionID;

/// Gets the function ID for a given SDL RPC function name
///
/// @param functionName The RPC function name
- (nullable NSNumber<SDLInt> *)functionIdForName:(SDLRPCFunctionName)functionName;

@end

NS_ASSUME_NONNULL_END
```

> Each <function> from MOBILE_API.XML is declares its function name in `SDLRPCFunctionNames.h` and `SDLRPCFunctionNames.m` files.

Reviewer 2: ```suggestion
Each <function> from MOBILE_API.XML declares its function name in `SDLRPCFunctionNames.h` and `SDLRPCFunctionNames.m` files.
```

```objc
SDLRPCFunctionNames.h
#import "SDLEnum.h"
/**
* All RPC request / response / notification names
*/
typedef SDLEnum SDLRPCFunctionName SDL_SWIFT_ENUM;

/// Function name for an AddCommand RPC
extern SDLRPCFunctionName const SDLRPCFunctionNameAddCommand;

/// Function name for an AddSubMenu RPC
extern SDLRPCFunctionName const SDLRPCFunctionNameAddSubMenu;

. . . and so on
```

And the implementation file SDLRPCFunctionNames.m :

```objc
//
//  SDLRPCFunctionNames.m
//  SmartDeviceLink
//

#import "SDLRPCFunctionNames.h"

SDLRPCFunctionName const SDLRPCFunctionNameAddCommand = @"AddCommand";
SDLRPCFunctionName const SDLRPCFunctionNameAddSubMenu = @"AddSubMenu";

. . . and so on
```

> 

Reviewer 2: This is missing information on `SDLRPCParameterNames`

Reviewer: please have look into https://github.com/smartdevicelink/sdl_ios/pull/1556/commits/4b7197ee027bb8b437a427cfdf3bbca01c6cca02

##### Request Functions (SDLRPCRequest)

> ##### Function with simple params

Reviewer 2: ```suggestion
```

> This section depicts all functions which include one or a few parameters of the following types (integer|decimal|boolean|string). Such parameters are considered as simple.

Reviewer 2: ```suggestion
```

```xml
    <function name="GetCloudAppProperties" functionID="GetCloudAppPropertiesID" messagetype="request" since="5.1">
        <description>
            RPC used to get the current properties of a cloud application
        </description> 
        <param name="appID" type="String" maxlength="100" mandatory="true"></param>
    </function>
```

```objc
//  SDLGetCloudAppProperties.h
//

#import "SDLRPCRequest.h"

NS_ASSUME_NONNULL_BEGIN

/**
 * RPC used to get the current properties of a cloud application
 *
 * @since SDL 5.1.0
 */
@interface SDLGetCloudAppProperties : SDLRPCRequest

/**
 * @param appID
 * @return A SDLGetCloudAppProperties object
 */
- (instancetype)initWithAppID:(NSString *)appID;

/**
 * {"default_value": null, "max_length": 100, "min_length": 1}
 *
 * Required, NSString *
 */
@property (strong, nonatomic) NSString *appID;

@end

NS_ASSUME_NONNULL_END
```

> 

Reviewer 2: ```suggestion
```

```objc
//  SDLGetCloudAppProperties.m
//

#import "SDLGetCloudAppProperties.h"
#import "NSMutableDictionary+Store.h"
#import "SDLRPCFunctionNames.h"

NS_ASSUME_NONNULL_BEGIN

@implementation SDLGetCloudAppProperties

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
- (instancetype)init {
    self = [super initWithName:SDLRPCFunctionNameGetCloudAppProperties];
    if (!self) { return nil; }

    return self;
}
#pragma clang diagnostic pop

- (void)setAppID:(NSString *)appID {
    [self.parameters sdl_setObject:appID forName:SDLRPCParameterNameAppId];
}

- (NSString *)appID {
    NSError *error = nil;
    return [self.parameters sdl_objectForName:SDLRPCParameterNameAppId ofClass:NSString.class error:&error];
}

@end

NS_ASSUME_NONNULL_END
```

##### Response Functions (SDLRPCResponse)

> ##### Response class with extra argument example (Alert Response):

Reviewer 2: ```suggestion
```

	</param>

	<param name="resultCode" type="Result" platform="documentation" mandatory="true">
		<description>See Result</description>
		<element name="SUCCESS"/>
		<element name="INVALID_DATA"/>
		<element name="OUT_OF_MEMORY"/>
		<element name="TOO_MANY_PENDING_REQUESTS"/>
		<element name="APPLICATION_NOT_REGISTERED"/>
		<element name="GENERIC_ERROR"/>
		<element name="REJECTED"/>
		<element name="ABORTED"/>
		<element name="DISALLOWED"/>
		<element name="USER_DISALLOWED"/>
		<element name="UNSUPPORTED_RESOURCE"/>
		<element name="WARNINGS"/>
	</param>

	<param name="info" type="String" maxlength="1000" mandatory="false" platform="documentation">
		<description>Provides additional human readable info regarding the result.</description>
	</param>

	<param name="tryAgainTime" type="Integer" minvalue="0" maxvalue="2000000000" mandatory="false" since="2.0">
		<description>
			Amount of time (in seconds) that an app must wait before resending an alert.
			If provided, another system event or overlay currently has a higher priority than this alert.
			An app must not send an alert without waiting at least the amount of time dictated.
		</description>
	</param>
</function>
```

> As we can see it in the XML it requires an additional param `tryAgainTime` which is declared and implemented in the `SDLAlertResponse` class as follows:

Reviewer 2: ```suggestion
```

```objc
//  SDLAlertResponse.h
//

#import "SDLRPCResponse.h"

NS_ASSUME_NONNULL_BEGIN

/**
 * @since SDL 1.0.0
 */
@interface SDLAlertResponse : SDLRPCResponse

/**
 * @param @(tryAgainTime)
 * @return A SDLAlertResponse object
 */

> - (instancetype)initWithTryAgainTime:(UInt32)tryAgainTime;

Reviewer 2: I know this is complicated, but I think that `init`s including optional parameters (like `tryAgainTime` is) should use the `NSNumber` version of the parameter and make it `nullable`. Mandatory parameters can use `UInt32` since they're not optional anyway. Would that be difficult to implement?

Reviewer: done in https://github.com/smartdevicelink/sdl_ios/pull/1556/commits/88a2ae6d3ba1d92c3444893a7509078f5da1ac5e
```objc
- (instancetype)initWithTryAgainTime:(nullable NSNumber<SDLUInt> *)tryAgainTime;
```

 * the amount of time dictated.
 * {"default_value": null, "max_value": 2000000000, "min_value": 0}
 *
 * @since SDL 2.0.0
 *
 * Optional, UInt32
 */
@property (nullable, strong, nonatomic) NSNumber<SDLUInt> *tryAgainTime;

@end

NS_ASSUME_NONNULL_END
```

```objc
//  SDLAlertResponse.m
//

#import "SDLAlertResponse.h"
#import "NSMutableDictionary+Store.h"
#import "SDLRPCFunctionNames.h"

NS_ASSUME_NONNULL_BEGIN

@implementation SDLAlertResponse

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
- (instancetype)init {
    self = [super initWithName:SDLRPCFunctionNameShowAppMenu];
    if (!self) { return nil; }

    return self;
}
#pragma clang diagnostic pop

- (void)setTryAgainTime:(nullable NSNumber<SDLUInt> *)tryAgainTime {
    [self.parameters sdl_setObject:tryAgainTime forName:SDLRPCParameterNameTryAgainTime];
}

- (nullable NSNumber<SDLUInt> *)tryAgainTime {
    return [self.parameters sdl_objectForName:SDLRPCParameterNameTryAgainTime ofClass:NSNumber.class error:nil];
}

@end

NS_ASSUME_NONNULL_END
```

##### Notification Functions (SDLRPCNotification)

> Another example is the class `SDLOnAppInterfaceUnregistered`. It is declared in the XML as follows:

Reviewer 2: ```suggestion
```

```xml
<function name="OnAppInterfaceUnregistered" functionID="OnAppInterfaceUnregisteredID" messagetype="notification" since="1.0">
	<param name="reason" type="AppInterfaceUnregisteredReason" mandatory="true">
	    <description>See AppInterfaceUnregisteredReason</description>
	</param>
</function>
```

> The corresponding Objective-C class `SDLOnAppInterfaceUnregistered` looks as following.

Reviewer 2: ```suggestion
```

```objc
//  SDLOnAppInterfaceUnregistered.h

#import "SDLRPCNotification.h"

@class SDLAppInterfaceUnregisteredReason;

NS_ASSUME_NONNULL_BEGIN

/**
 * @since SDL 1.0.0
 */
@interface SDLOnAppInterfaceUnregistered : SDLRPCNotification

/**
 * @param reason - @(reason)
 * @return A SDLOnAppInterfaceUnregistered object
 */
- (instancetype)initWithReason:(SDLAppInterfaceUnregisteredReason)reason;

/**
 * See AppInterfaceUnregisteredReason
 *
 * Required, SDLAppInterfaceUnregisteredReason 
 */
@property (strong, nonatomic) SDLAppInterfaceUnregisteredReason reason;

@end

NS_ASSUME_NONNULL_END
```

> *Note: though the `SDLAppInterfaceUnregisteredReason reason` declared as a property but implemented as custom getter and setter*

Reviewer 2: ```suggestion
```

```objc
//  SDLOnAppInterfaceUnregistered.m

#import "SDLOnAppInterfaceUnregistered.h"
#import "NSMutableDictionary+Store.h"
#import "SDLAppInterfaceUnregisteredReason.h"
#import "SDLRPCParameterNames.h"
#import "SDLRPCFunctionNames.h"

NS_ASSUME_NONNULL_BEGIN

@implementation SDLOnAppInterfaceUnregistered

#pragma clang diagnostic push

> #pragma clang diagnostic ignored "-Wdeprecated-declarations"

Reviewer 2: Notification inits are currently deprecated, but they should not be. Can we generate them as not deprecated please?

Reviewer: Do you talking about this lines:
```objc
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
... 
#pragma clang diagnostic pop
```
@joeljfischer the same lines are present in all already existing `functions`, do you want to delete this lines from template?

Reviewer 2: I apologize, that was a mistake on my part. I looked back into it again to confirm, and you're correct, they should not be deleted.

  -xml SOURCE_XML, --source-xml SOURCE_XML, --input-file SOURCE_XML
                        should point to MOBILE_API.xml
  -xsd SOURCE_XSD, --source-xsd SOURCE_XSD
  -d OUTPUT_DIRECTORY, --output-directory OUTPUT_DIRECTORY
                        define the place where the generated output should be
                        placed
  -t [TEMPLATES_DIRECTORY], --templates-directory [TEMPLATES_DIRECTORY]
                        path to directory with templates
  -r REGEX_PATTERN, --regex-pattern REGEX_PATTERN
                        only elements matched with defined regex pattern will
                        be parsed and generated
  --verbose             display additional details like logs etc
  -e, --enums           if present, all enums will be generated
  -s, --structs         if present, all structs will be generated
  -m, -f, --functions   if present, all functions will be generated
  -y, --overwrite       force overwriting of existing files in output
                        directory, ignore confirmation message
  -n, --skip            skip overwriting of existing files in output
                        directory, ignore confirmation message
```

> ### How to use the generated classes

Reviewer 2: Can you add a section here on how to run the unit tests?

Reviewer: done in https://github.com/smartdevicelink/sdl_ios/pull/1556/commits/a179e65851c4c239d5bef67a0f3ca591dd235dec
